File: packages/app/src/services/software/definitions/elgatoStreamDeck.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'elgato-stream-deck';
	public name = 'Elgato Stream Deck';
	public category = [SoftwareCategory.Peripheral];
	public downloadName = 'Stream_Deck.msi';
	public shouldCacheUrl = true;
	public icon = 'elgato-stream-deck.png';
	public homepage = 'https://help.elgato.com/hc/en-us/sections/5162671529357-Elgato-Stream-Deck-Software-Release-Notes';

	public async resolveDownloadUrl() {
		const res = await fetch('https://elgato.com/us/en/s/downloads');
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const html  = await res.text();
		const match = html.match(/https:\/\/edge\.elgato\.com\/egc\/windows\/sd\/Stream_Deck_\d+\.\d+\.\d+\.\d+\.msi/);
		if (!match) {
			return err(DownloadUrlResolveError.Generic);
		}

		return ok(match[0]);
	}
}

File: src/lib/software/elgato_stream_deck.py
from re import compile
from PySide6.QtCore import Slot
from src.lib.software import BaseSoftware, SoftwareCategory
from PySide6.QtNetwork import QNetworkReply, QNetworkRequest

class ElgatoStreamDeck(BaseSoftware):
    def __init__(self):
        super().__init__()

        self.key = 'elgato-stream-deck'
        self.name = 'Elgato Stream Deck'
        self.category = [SoftwareCategory.Peripheral]
        self.download_name = 'Stream_Deck.msi'
        self.should_cache_url = True
        self.icon = 'elgato-stream-deck.png'
        self.homepage = 'https://help.elgato.com/hc/en-us/sections/5162671529357-Elgato-Stream-Deck-Software-Release-Notes'

    @Slot(QNetworkReply)
    def on_manager_finished(self, reply: QNetworkReply):
        reply.deleteLater()
        error = reply.error()
        if error != QNetworkReply.NetworkError.NoError:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            return

        download_url_pattern = compile(r'https://edge\.elgato\.com/egc/windows/sd/Stream_Deck_\d+\.\d+\.\d+\.\d+\.msi')
        html = reply.readAll().data().decode()
        match = download_url_pattern.search(html)
        if not match:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
        else:
            self.url_resolved.emit(match.group(0))

    def resolve_download_url(self):
        self.manager.get(QNetworkRequest('https://www.elgato.com/us/en/s/downloads'))
